### ISSUE TITLE
AWSCognitoAuth callback is not executed when login is canceled

### ISSUE THREAD

**Issue by takumi-sria** [aws-amplify/aws-sdk-ios#1043]
**Describe the bug**
* Callback is not executed when login is canceled
* Can not get error when login canceled
* I can not get error when cancellation after getSession execution

**To Reproduce**
1. Start the app and call getSession.
2. Tap "Done" on the login screen to cancel login.
3. AWSCognitoAuthGetSessionBlock: callback not invoked

```
let auth: AWSCognitoAuth = AWSCognitoAuth.default()
var session: AWSCognitoAuthUserSession?
....

self.auth.getSession  { (session:AWSCognitoAuthUserSession?, error:Error?) in
    //callback not invoked
    if((error as NSError).code == -1000){
        print("User canceled")
    }else {
      self.session = session
    }
}
```

**Which AWS service(s) are affected?**
AWSCognitoAuth

**Expected behavior**
1. Tap "Done" on the login screen to cancel login
2. AWSCognitoAuthGetSessionBlock: callback is called
3. Error comes back in return value

**Environment(please complete the following information):**
 - SDK Version: 2.6.29
 - Dependency Manager:Cocoapods
 - Swift Version : 4.1.2

**Device Information (please complete the following information):**
 - Device: iPhone6, Simulator
 - iOS Version: iOS 9.3,iOS 11

**Additional context**
AWS Cognito Auth (2.6.28) can successfully obtain an error at the time of cancellation execution.

**Comment by JesusMartinAlonso** [comment]
I have the same problem. I have debugged the SDK and the problems appears to be in the ```-(void) dismissSafariVC: (void (^)(void)) dismissBlock``` function. The ``` [self.svc dismissViewControllerAnimated:NO completion:^{``` works correctly but the completion is not getting called, so the completion of ```getSession``` is never called. This is the whole code of the function:
```

**
 Dismiss and reap the safari view controller
 */
-(void) dismissSafariVC: (void (^)(void)) dismissBlock {
    dispatch_async(dispatch_get_main_queue(), ^{
        if(self.svc){
            [self.svc dismissViewControllerAnimated:NO completion:^{
                //clean up vc
                self.svc = nil;
                dismissBlock();
            }];
        }else {
            dismissBlock();
        }
    });
}
```

**Comment by omerwaqaskhan** [comment]
Facing the same issue on, not getting the session after signing in 

```
auth.getSession(self) { (session:AWSCognitoAuthUserSession?, error:Error?) in
            
            if error != nil {
                print((error! as NSError).userInfo["error"] as? String ?? "")
            }
            else {
                self.session = session
            }
            self.refresh()
        }
```

**Comment by omerwaqaskhan** [comment]
Any solution found to the problem?

**Comment by omerwaqaskhan** [comment]
Any update?

**Comment by joelmbell** [comment]
Have encountered a similar issue as well. Looks callback to the main queue doesn't get executed sometimes.

```
 dispatch_async(dispatch_get_main_queue(), ^{
     /// this never executes
 }
```

**Comment by omerwaqaskhan** [comment]
Any solution?

**Comment by omerwaqaskhan** [comment]
Is there any solution to it ?

**Comment by rohandubal** [comment]
Hello @omerwaqaskhan 

We will look into this issue soon. 

I would like to point that we released a new SDK (2.6.35+) which now supports SFAuthenticationSession which is the recommended way of handling Web Authentication. Could you please try 2.6.35 and tell if it resolves your issue?

In the meanwhile, we will try to reproduce this on our side. 

Thanks,
Rohan

**Comment by rohandubal** [comment]
Sorry, closed by mistake. Re-opened.

**Comment by omerwaqaskhan** [comment]
> Hello @omerwaqaskhan
> 
> We will look into this issue soon.
> 
> I would like to point that we released a new SDK (2.6.35+) which now supports SFAuthenticationSession which is the recommended way of handling Web Authentication. Could you please try 2.6.35 and tell if it resolves your issue?
> 
> In the meanwhile, we will try to reproduce this on our side.
> 
> Thanks,
> Rohan

Hi @rohandubal 

Sure I will look in to SFAuthenticationSession but could you please confirm if it goes smooth with Cognito?

Thanks 
Omer

**Comment by rohandubal** [comment]
@takumi-sria @joelmbell  @omerwaqaskhan  This issue is resolved in SDK version `2.7.2` 

Please try it and let us know if you face any further issues.

Thanks,
Rohan
